File: app/config/storage.py
from google.cloud import storage
from app.config.settings import settings
from app.utils.logger import logger


class GoogleCloudStorage:
    _instance = None
    _client: storage.Client = None
    _bucket: storage.Bucket = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(GoogleCloudStorage, cls).__new__(cls)
        return cls._instance

    def __init__(self):
        if not self._client:
            try:
                self._client = storage.Client(
                    project=settings.GCP_PROJECT_ID
                )
                self._bucket = self._client.bucket(settings.GCP_BUCKET_NAME)
                logger.info("Google Cloud Storage initialized")
            except Exception as e:
                logger.warning(f"GCP initialization error: {e}")
                logger.warning("Google Cloud Storage features will be limited")

    def get_bucket(self) -> storage.Bucket:
        """Get GCS bucket"""
        return self._bucket

    def get_client(self) -> storage.Client:
        """Get GCS client"""
        return self._client


gcs = GoogleCloudStorage()

File: backend/src/config/storage.ts
import { Storage } from '@google-cloud/storage';
import logger from '../utils/logger';

let storage: Storage;

try {
  const projectId = process.env.GCP_PROJECT_ID;
  const keyFilename = process.env.GOOGLE_APPLICATION_CREDENTIALS;

  if (!projectId) {
    logger.warn('GCP_PROJECT_ID not set. Google Cloud Storage features will be limited.');
  }

  storage = new Storage({
    projectId,
    keyFilename,
  });

  logger.info('Google Cloud Storage initialized');
} catch (error) {
  logger.error('Failed to initialize Google Cloud Storage:', error);
  throw error;
}

export const bucketName = process.env.GCP_BUCKET_NAME || 'shorts-downloader-temp';

export default storage;
